### PR TITLE
fix: ignore existing target status when adding new targets

### PR DESCRIPTION
**Pull request by jiuker**
## Description

ignore exsits target status when add/set/delete

## Types of changes
- [x] Bug fix [#17147 ](https://github.com/minio/minio/issues/17147)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by shtripat** [comment]
Tested the changes. It works if old web-hook if offline. But allows setting a new one which is down as well. I understand that's acceptable as its a configuration and runtime while sending the audit, it would fail if target not reachable.
With this assumption, change looks good to me.

**Review by shtripat — APPROVED** [review]
lgtm

**Comment by harshavardhana** [comment]
This actually reminds me that we have to the client CLI UI slightly differently here, because we have always expected targets to be online when they are being set.

Basically, we want a solution like this 

- Allow adding new targets while having some existing offline targets
- Allow deletes to happen on any target
- However do not allow offline targets to be added freshly, we want
  them to be online when they are being added - just doesn't make
  sense to add a config that is not valid. 

In this PR the 3rd part is not complete yet.

**Comment by jiuker** [comment]
@harshavardhana review again

**Comment by jiuker** [comment]
@shtripat make a test ,plz

**Comment by shtripat** [comment]
> @shtripat make a test ,plz

How is third point from @harshavardhana being taken care? I am still able to add a target which is offline as below
```
$ mc admin config set ALIAS notify_webhook:3 enable=on endpoint=http://localhost:8000/webhook
Successfully applied new settings.
Please restart your server 'mc admin service restart myminio'.
```
Here `http://localhost:8000/webhook` is not actually up.

Am I missing something?

**Comment by jiuker** [comment]
> > @shtripat make a test ,plz
> 
> How is third point from @harshavardhana being taken care? I am still able to add a target which is offline as below
> 
> ```
> $ mc admin config set ALIAS notify_webhook:3 enable=on endpoint=http://localhost:8000/webhook
> Successfully applied new settings.
> Please restart your server 'mc admin service restart myminio'.
> ```
> 
> Here `http://localhost:8000/webhook` is not actually up.
> 
> Am I missing something?

No，I don't use mc.Thx，I will test again.

**Comment by jiuker** [comment]
@shtripat It's ready.Make a test.Plz

**Comment by shtripat** [comment]
> @shtripat It's ready.Make a test.Plz

There seems to different issue now as unable to set `notify_webhook` value

```
$ mc admin config set myminio notify_webhook enable=on endpoint=http://localhost:8000/webhook
Successfully applied new settings.
Please restart your server 'mc admin service restart myminio'.

$ mc admin service restart myminio 
Restart command successfully sent to `myminio`. Type Ctrl-C to quit or wait to follow the status of the restart process.
...
Restarted `myminio` successfully in 1 seconds

$ mc admin config get myminio audit_webhook
audit_webhook enable=off endpoint= auth_token= client_cert= client_key= queue_size=100000 queue_dir= 
```

Webhook is actually up and able to receive requests

```
$ curl -XPOST -H 'Content-Type: application/json' -d '{"msg": "hello"}' http://localhost:8000/webhook
Webhook received!
```

**Comment by jiuker** [comment]
> > @shtripat It's ready.Make a test.Plz
> 
> There seems to different issue now as unable to set `notify_webhook` value
> 
> ```
> $ mc admin config set myminio notify_webhook enable=on endpoint=http://localhost:8000/webhook
> Successfully applied new settings.
> Please restart your server 'mc admin service restart myminio'.
> 
> $ mc admin service restart myminio 
> Restart command successfully sent to `myminio`. Type Ctrl-C to quit or wait to follow the status of the restart process.
> ...
> Restarted `myminio` successfully in 1 seconds
> 
> $ mc admin config get myminio audit_webhook
> audit_webhook enable=off endpoint= auth_token= client_cert= client_key= queue_size=100000 queue_dir= 
> ```
> 
> Webhook is actually up and able to receive requests
> 
> ```
> $ curl -XPOST -H 'Content-Type: application/json' -d '{"msg": "hello"}' http://localhost:8000/webhook
> Webhook received!
> ```

@shtripat you enable audit_webhook to off? this is part of test?

**Comment by shtripat** [comment]
Oh that seems to be spell mistake from my side. Will re-verify..

**Comment by shtripat** [comment]
> Oh that seems to be spell mistake from my side. Will re-verify..

Sorry for type. It works now and doesn't allow to set one with down state. Also if one of targets is down, it does allow to set one which is up and running.

**Comment by jiuker** [comment]
> > Oh that seems to be spell mistake from my side. Will re-verify..
> 
> Sorry for type. It works now and doesn't allow to set one with down state. Also if one of targets is down, it does allow to set one which is up and running.

@shtripat thanks a lot

**Comment by harshavardhana** [comment]
PTAL @shtripat @Praveenrajmani

**Comment by Praveenrajmani** [comment]
can you fix the lint failure? @jiuker

**Review by shtripat — APPROVED** [review]
Other than lint issues, code looks good to me.
Please fix them.

**Comment by jiuker** [comment]
@Praveenrajmani run pipeline agagin.plz

**Review by shtripat — APPROVED** [review]
lgtm

**Review by harshavardhana — APPROVED** [review]

File: cmd/admin-handlers-config-kv.go [hunk 1]
```diff
@@ -211,7 +211,12 @@ func setConfigKV(ctx context.Context, objectAPI ObjectLayer, kvBytes []byte) (re
 		return
 	}
 
-	if verr := validateConfig(result.Cfg, result.SubSys); verr != nil {
+	tgts, err := config.ParseConfigTargetID(bytes.NewReader(kvBytes))
```
Comment (Praveenrajmani) [inline cmd/admin-handlers-config-kv.go:214]:
```go
result.SubSys, _, target, err = config.GetSubSys(string(kvBytes))
```

^^ already provides the target which is going to be edited/added. We can use this instead.

File: internal/config/config.go [hunk 2]
```diff
@@ -426,6 +426,34 @@ func (c Config) DelFrom(r io.Reader) error {
...
 }
 
+// ContextKeyString is type(string) for contextKey
+type ContextKeyString string
+
+// ContextKeyForTargetFromConfig - key for context for target from config
+const ContextKeyForTargetFromConfig = ContextKeyString("ContextKeyForTargetFromConfig")
+
+// ParseConfigTargetID - read all targetIDs from reader
+func ParseConfigTargetID(r io.Reader) (ids map[string]bool, err error) {
```
Comment (Praveenrajmani) [inline internal/config/config.go:436]: I don't think SetConfigKV API will accept multiple targets. It works with only single target. We don't need this parser.

File: internal/config/notify/parse.go [hunk 3]
```diff
@@ -58,13 +58,24 @@ func TestSubSysNotificationTargets(ctx context.Context, cfg config.Config, subSy
...
+			if err == nil && !yes {
+				err = ErrTargetsOffline
+			}
+			if err != nil {
+				return fmt.Errorf("error (%s): %w", target.ID(), err)
+			}
+		} else {
+			// Just for call init.
+			// Ignore target is online or offline
+			_, _ = target.IsActive()
```
Comment (Praveenrajmani) [inline internal/config/notify/parse.go:78]: IMO, We don't need to ping other targets too. This is a side effect and can be avoided. 

`Set|Delete` don't need to care about Init.

Need your help here to decide @harshavardhana

**Review by Praveenrajmani — COMMENTED** [review]
Thanks for the PR @jiuker 

Just some minor things to check.

File: cmd/admin-handlers-config-kv.go (hunk 1 above)
Comment (jiuker) [reply cmd/admin-handlers-config-kv.go:214]: @Praveenrajmani It used in ParseConfigTargetID

File: internal/config/config.go (hunk 2 above)
Comment (jiuker) [reply internal/config/config.go:436]: @harshavardhana help there too.

**Pull request merged**